flake.nix

```diff
@@ -0,0 +1,41 @@
+{
+  description = "FormalSystems lean implementation";
+
+  inputs = {
+    lean.url = "github:leanprover/lean4/v4.2.0";
+    flake-utils.url = "github:numtide/flake-utils";
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05";
+  };
+
+  outputs = { self, lean, flake-utils, nixpkgs }: flake-utils.lib.eachDefaultSystem (system:
+    let
+      leanPkgs = lean.packages.${system};
+      pkgs = import nixpkgs { system = system; };
+      leanDeps = map
+        (dep: {
+          path = builtins.fetchGit {
+            inherit (dep.git) url rev name;
+          };
+          inherit (dep.git) name;
+        })
+        (builtins.fromJSON (builtins.readFile ./lake-manifest.json)).packages;
+      pkg = pkgs.stdenv.mkDerivation {
+        name = "formal-systems";
+        src = ./.;
+        nativeBuildInputs = [ leanPkgs pkgs.git ];
+        buildPhase = ''
+          ln -sf ${leanPkgs.lean-all}/* .
+          mkdir -p ./lake-packages
+          ${builtins.concatStringsSep "&&" (map (dep: "cp -r " + dep.path + " ./lake-packages/" + dep.name) leanDeps)}
+          bin/lake --version
+          bin/lake build
```

Comment:
@mmarx ok, so I now tried to manually fetch the deps and invoke lake, but it does not work, because lake expects the folders to be git repositories, and tries to download them again, which obviously fails. Is there some way to tell fetchGit to keep the .git folder?

Reply:
The `fetchgit` from `nixpkgs` has `leaveDotGit` (that seems to have its own reproducibility issues). Instead of using a custom `buildPhase`, can't you do something like this:
```nix
pkg = let deps = builtins.listToAttrs (map dep: 
  { inherit (dep.git) name; 
    value = leanPkgs.buildLeanPackage { 
      src = builtins.fetchGit { inherit (dep.git) url rev name; }; 
      inherit (dep.git) name; 
    };
  });
in leanPkgs.buildLeanPackage {
  src = ./.;
  deps = [leanPkgs.lean] ++ (builtins.attrValues deps)
};
```

This will probably need some more work to handle transitive dependencies, but then you're not relying on having an actual git repository around.

Reply:
well I tried that, but I don't understand what kind of objects is expected in `deps`, at least this version does not work, because somewhere inside the `buildLeanPackage` code it tries to read non-present attributes. Also `buildLeanPackage` expects the package name to be the same as the top-level .lean file, which already breaks down for mathlib.

Reply:
Yeah, turned out this was not _so_ trivial to do, but this seems to work (except that it actually doesn't build because of a “simp made no progress” in FormalSystems/Preliminaries/Language.lean).